src/main/java/htsjdk/samtools/cram/build/Sam2CramRecordFactory.java

```diff
@@ -291,36 +271,62 @@ private void addInsertion(final List<ReadFeature> features, final int zeroBasedP
         }
     }
 
-    private void addSubstitutionsAndMaskedBases(final CramCompressionRecord cramRecord, final List<ReadFeature> features, final int fromPosInRead, final int
+    /**
+     * Processes a stretch of read bases marked as match or mismatch and emits appropriate read features.
+     * Briefly the algorithm is:
+     * <ul><li>emit nothing for a read base matching corresponding reference base.</li>
+     * <li>emit a {@link Substitution} read feature for each ACTGN-ACTGN mismatch.</li>
+     * <li>emit {@link ReadBase} for a non-ACTGN mismatch. The side effect is the quality score stored twice.</li>
+     * <p>
+     * The rest of the method handles quality score read features added if refSNPs is specified.
+     * <p>
+     * IMPORTANT: reference and read bases are always compared for match/mismatch in upper case due to BAM limitations.
+     *
+     * @param cramRecord           a record to work on
+     * @param features             a list of read features to add to
```

Comment:
I meant that it could return the list *instead* of having it as an argument. Then it would be clear from the signature and the call site that the list is an output, not an input. Also, the resulting method would then be a pure function with no side-effects. But it was just a suggestion...